### ISSUE TITLE
[RFC] TechdocsX (MDX support in techdocs)

### ISSUE THREAD

**Issue by joaomilho** [backstage/backstage#4548]
**Status:** Open for comments
Given my lack of context, please point out anything that I might miss or have gotten wrong.

## Need

In its current form, techdocs allows markdown and it relies on plugins to add extra functionality. [MDX](https://github.com/mdx-js/mdx) has been growing in popularity in the past few years and would allow enriching our docs experience in a developer friendly way.

The rationale is as follows:

1. _The augmented hero's journey_
Stories are more personal and engaging when interactive. Following the dawn of the interactive [news stories](https://www.nytimes.com/interactive/2020/12/30/us/2020-year-in-graphics.html) , perhaps the time has come for ~docs~ [stories](https://observablehq.com/@jashkenas/inputs) [as well](https://observablehq.com/@observablehq/five-minute-introduction).

2. _React is ubiquitous_
MDX is simply React. All developers with React knowledge would be able to contribute with extra functionality. Additionally, markdown plugins add ad-hoc syntax that is largely unknown to developers and hard to teach through documentation. With MDX, the syntax would be the same for all extensions. Additionally, documentation for the available components can be embedded in the docs, themselves, with… MDX.

3. _Writing MDX components is easier than integrating MD plugins_
To integrate plugins today, one has to extend the core functionality of techdocs. With MDX the available components can be externalized, both by providing a standard library and allowing ad-hoc contributions for 3rd party Backstage installations (see proposed solution below).

4. _Innovation speed and wow effect_
The point of this proposal is not to list all things we could do with MDX (although a modest list is provided below), but point to the fact that once this is enabled, we could expect increased and faster contributions from both internal and external users. A great step in becoming a leader in developer experience.

## Proposal

To integrate MDX with our current offering successfully I envision the following requirements (please comment if there are important requirements missing):

- Be as backwards compatible with the current implementation as possible.
- To be able to render MDX remotely and in a safe way. Ideally, we shouldn’t move any javascript down the wire and do not rely on any evaluation of code.
- Ensure the components follow design guidelines.
- Document the feature extensively.

The proposed solution to the above is as follows:

1. _Standard library_
Create a standard library for the available components (say @backstage/mdx-components). These are maintained by us and follow our design guideline (requirement 3).

2. _Backward compatibility_
When creating a project, developers can still use only MD. Since MDX is mainly a superset of MD, these would be consumed almost the same way they are today (requirement 1). A bit of extra research would be needed to map out all the breaking changes, but they are known to exist mainly when rendering HTML inside MD and this is not a particularly hard problem to solve.

3. _Incremental adoption_
Developers can opt in to use MDX in their repos. The way they would do that would be by installing our component library and registering the components implicitly (as in the components property of the MDX provider) in their MDX renderer (that is, no import syntax). Since we would only allow for these components, no ad-hocs components written in the MDX files would work. Locally, since the components are available docs would simply work and render the components correctly.

4. _Security_
The MDX content in these files would move down the wire in their text representation. That is, not as JSON or anything fancy, literally as, e.g. `<PieChart data={[{mdx: 70, md: 30}]} />` (requirement 2).

5. _Rendering_
The rendering side (aka Backstage) would then interpret the text representation and render the components from the same library. This is already done in the wild, for instance with the [next-mdx-remote](https://github.com/hashicorp/next-mdx-remote) package provided by [Hashicorp](https://www.hashicorp.com/). Although the name mentions nextjs, the library can actually be used elsewhere and includes functionality for server side rendering, client side “[hydration](https://medium.com/@luke_schmuke/how-we-achieved-the-best-web-performance-with-partial-hydration-20fab9c808d5#:~:text=SSR%20and%20hydration&text=This%20means%20that%20Next%20will,HTML%20using%20your%20React%20code)”, context providers (so that components can have shared state and additional functionality) and more.

6. _Documentation_
The standard lib can be documented in MDX itself and display its components, usage, types and props, how to contribute, etc (requirement 4).

## Possibilities
An image is worth a thousand words, so here are some examples of things that MDX could enable:

- Components with state.
  - Variables in docs, as in {your-service-name}. With MDX one could ask the user to enter the service name once and update all the instances of the variables in the docs. Or these variables could even be taken from the user context (say {your-username}) or be remotely fetched. One interesting example of such a thing are the dropdowns in [asdf docs](https://asdf-vm.com/#/core-manage-asdf) and others. <br /><img width="350" alt="asdf docs select example" src="https://user-images.githubusercontent.com/308196/108067899-b5d97c80-7061-11eb-8871-6a8d607a42f0.png">
  - Allow users to write personal comments, highlight or pin content to read later like in [pocket](https://help.getpocket.com/article/1144-highlighting-in-pocket). <img alt="highlight content with pocket" src="https://d33v4339jhl8k0.cloudfront.net/docs/assets/56bbd994c697917fdc55e05e/images/5ad67a672c7d3a0e93676cd6/file-eLRsFmOKSC.png" width="350" />

- Simple ad-hoc forms that today would require contribution to Backstage. One could build polls, feedback forms, quizzes that could be integrated in instructional material like in [NextJS docs](https://nextjs.org/learn/basics/navigate-between-pages/pages-in-nextjs), etc. <img width="350" alt="nextjs docs quizz" src="https://user-images.githubusercontent.com/308196/108068583-8d9e4d80-7062-11eb-88d0-a32c88f454cb.png">

- Embedding of different file formats, beyond graphviz.

- Graphs & charts and dataviz that could use local or remote data like in [nevo docs](https://nivo.rocks/bar/). <img width="350" alt="Screenshot 2021-02-16 at 14 39 00" src="https://user-images.githubusercontent.com/308196/108070631-07cfd180-7065-11eb-88d6-780eb90f70f4.png">

- [Live coding](https://github.com/FormidableLabs/react-live) for instructional docs or even more ambitious [sketchbooks](http://worrydream.com/#!2/LadderOfAbstraction) like in [observablehq](http://observablehq). <img src="https://user-images.githubusercontent.com/17658189/63181897-1d67d380-c049-11e9-9dd2-7da2a3a57f05.gif" width="350" />

- Enhanced functionality for code blocks, like smart copy/paste for diffs, link to remote, auto GH search... Code blocks and highlight could be handled by MDX instead of rendered previously, allowing for more languages faster and other things.

## Risks, downsides, other thoughts

- Versioning of the standard libraries. That is, if a doc uses an incompatible version of a component it could break the renderer. A problem to solve.

- MDX can't be immediately previewed at github, github enterprise, bitbucket and most other systems with some MD support since it relies on extra compilation steps. The MD parts of it are still visible 'tho.

- Backward compatibility could prove to be a bit harder than anticipated, given the current dependency on mkdocs and the interaction with plain HTML in MD.

- Indexing of MDX content.

- Alienating users. If a teams with both people who can write React and people who can relies on MDX too much it may become a barrier to contribute from those who don't, although I imagine teams mainly falling into 2 buckets: the ones who simply use MD and the ones who leverage MDX and most members can maintain it.

- There's a need for teams to publish their own component libraries docs. Given this would involve ad-hoc components, this is not detailed here, but a similar solution apart from techdocs could be considered.

- It is conceivable that for many simpler use cases, MDX docs would be a simpler solution than writing backstage plugins themselves. I can foresee developers writing simple components as part of their docs instead of writing and integrating a full blown plugin.

- A future Backstage desktop app could interact with the standard components, and given proper permissions, perform automation tasks, like running a command automatically in the terminal (waiting for its conclusion and checking the tutorial step), opening an app and performing a task, etc.

**Comment by joaomilho** [comment]
A little MVP can be found here https://github.com/joaomilho/backstage/tree/mvp-techdocs-next (diff: https://github.com/backstage/backstage/compare/master...joaomilho:mvp-techdocs-next?expand=1)

To run it one must replace the github url and key, I didn't have time to figure how to handle config yet 🤷

**Comment by iamEAP** [comment]
Hey @joaomilho!  Love the thinking here, especially through the lens of, "it is difficult to extend TechDocs to fit my custom needs," which we've occasionally seen internally.  Writing mkdocs plugins (or even just picking existing ones off the shelf and incorporating them into the TechDocs build process) is not straightforward and can be limiting.

A couple of thoughts:

### On interdependency
Out of necessity (for similar security reasons as you outlined above), we've already gone part way down a path where we have interdependencies between TechDocs generated documentation and JavaScript in Backstage, which has resulted maintenance headaches.

In the current case, we have JS in Backstage that relies on specific markup (in order to re-implement some interactivity that gets stripped out for XSS mitigation reasons), but when the generated markup changes (underlying libraries in the build process evolve and the exact markup isn't part of their API), we have to update Backstage code to take it into account (and in a backwards compatible way, since docs are built and not guaranteed to rebuilt).

In this proposal, we may end up with `<PieChart data={[{mdx: 70, md: 30}]} />` in a documentation page that relies on an equivalent component in `@backstage/mdx-components`, but because those things are now distributed across repos (and in the case of a standard library, across Backstage sites and companies), evolution of the component becomes constrained and updating it becomes a complex undertaking.

### On React ubiquity
React is definitely popular, and JSX is definitely the most common way of authoring React components.  But I think it's important to consider TechDocs' audience (on the documentation authoring/publishing side of the platform): not every docs author is an engineer (let alone a front-end engineer, and let alone a front-end engineer who knows JSX).

Granted, MDX is a superset (so there would be nothing stopping someone from just continuing to use vanilla markdown to author their docs), this would inevitably create some friction for both **a)** non-engineers maintaining documentation (encountering a block of what looks like gibberish and not knowing what to do with it), but also **b)** engineers writing new documentation (who, either have to learn React/JSX, or who may already know it, but who now have to author it in a less-than-ideal development environment).

### Overall
I've definitely experienced some great, interactive, stateful documentation sites out in the wild, but I wonder if the use-cases for those are narrower than what TechDocs aims to solve, and maybe we can solve those particular use-cases more directly (e.g. a sister platform/plugin for amazing, interactive API docs), without a grand "JSX in MD" compromise.

**Comment by joaomilho** [comment]
So,

1) I appreciate the caution, I've written this RFC but I also think we should take things lightly and with the proper care.

2) Interdependency: Agreed. I completely missed "versioning" as a risk in the RFC. How come? It's probably the hardest problem to solve here indeed (updating the RFC now). And I honestly don't have a killer suggestion to solve it. Let's see if anyone reading this has suggestions :D

3) Ubiquity: Agreed. I also think that this friction may exist if docs are shared between tech/non-tech users, although I would add that more often than not MD itself is already a barrier (GDocs import would be the most general solution TBH). I would imagine most often the scenarios would look like either teams using just MD or teams using MDX where all or most members can maintain it, but there might be a good % in the wrong part of the Venn diagram.

4) Overall. Agreed. I think only a certain part of our audience would actually leverage this, even though this cases would bring a lot of value. We could try this as a different plugin and only consider embedding it in techdocs after some learnings.

**Comment by joaomilho** [comment]
Just a final remark on your comment, although I think the difficulty of extending MD as it is an important aspect of the RFC, I _stronly_ think we shouldn't see it as a solution to that problem in particular, but as an enabler of other things.

**Comment by Rugvip** [comment]
Wondering if a version where MDX for a specific set of known components is useful. The TechDocs frontend plugin would be configured with both built-in and custom components, and those can then be used in MDX. Gives us a way to provide more powerful doc features, and a way for integrators and organizations a way to power up their docs as well, but without the headaches of dynamically loading in things and the security issues that come with it.

The components you want to make available in MDX would probably be provided with some strict `prop-types` definitions, likely a subset that only allows primitive values.

Possibly also combine with a bit of data collection in the backend so provide data on what components and props are in use and by what docs, allowing an organization to detect broken docs and plan migrations.

**Comment by joaomilho** [comment]
@Rugvip I'm not sure I fully understand the difference between your suggestion and the original one.

On the data collection part, it could be done when we SSR these components (aka `renderToString`).

**Comment by Rugvip** [comment]
@joaomilho got the impression that components were remote as well, because I didn't see any in the example, but guess that's not the case then :grin:

**Comment by ottosichert** [comment]
Reopening this RFC to gather some similar work streams around this topic:
- https://github.com/backstage/backstage/pull/3842 (Rocdocs)
- https://github.com/backstage/backstage/issues/3998 (Minimal MkDocs theme)
- https://github.com/backstage/backstage/tree/techdocsx/milestone-0-poc
- (Let me know if you have more relevant items!)

---

### High level goals

- Improve documentation creation experience (through local and faster feedback loop)
- Maintain documentation consumption experience (still having pre-built and cacheable assets)
- Reduce architectural complexity (through reducing transformation layers and exiting the MkDocs ecosystem)

### Some history

> "The architecture is this way is because TechDocs was a wildly successful hack project built on best-of-breed tools of the time.  “What if we generated docs using a popular docs framework and just iframe’d the HTML into Backstage?”
> And then that architecture was slowly evolved to where it is now to work around the limitations of that initial hack (shadow DOM, react portals, various mkdocs plugins, etc)" - @iamEAP 

> "Back in the days (techdocs legacy) we just used an iframe to embed the generated html, so that was a quick solution to get things in Backstage. Then we realized that we wanted to support more and more features on top of techdocs, plugins, which was on of the reasons Bil spent some hackdays on a proof of concept for TechDocs Native (shadowDom etc).
> As mkdocs does gives us a lot out of the box, with the whole plugin/extension community, we have also been able to contribute to this (with for example mkdocs-monorepo-plugin. And all of these extensions/plugins we use are things are things we would have to develop ourselves if we do not use mkdocs, so something one have to thing about." - @emmaindal 

### Comparison

| Aspect        | MkDocs           | TechdocsX  |
| ------------- |-------------| -----|
| Architectural layers      | - Markdown files<br>- MkDocs transformation<br>- MkDocs plugins extras<br>- HTML persistence<br>- Pre-transformers<br>- DOM post-transformers|- Markdown files<br>- MDX parser<br>- _OPTIONAL: persist transpiled JS from MDX_<br>- Custom components|
| Conforming to React principles | ❌ (mutating DOM) | ✅  |
| Rendering on the fly (without waiting for builds) | ❌ | ✅ |
| Stateless rendering (without persistence) | ❌ | ✅ |
| Fast local feedback loop while writing docs / developing plugins | ❌ | ✅ (Hot Module Reloading) |
| Supports React DevTools | ❌ (due to Shadow root) | ✅ |
| Contribution likelihood | ⚠️ (should it be a MkDocs plugin or transformer? complex to understand) | ✅ (just write a React component) |
| Supports GitHub flavored markdown | ❌  | ✅ |
| Easy testability | ⚠️ (E2E necessary for some cases) | ✅ |
| Access to MkDocs plugin ecosystem | ✅ | ❌ |
| Access to React component ecosystem | ⚠️ (hackily possible with post-transformers) | ✅ |
| Easy Markdown syntax | ✅ (although each plugin might have arbitrary syntax) | ⚠️ (JSX not easiest to pick up) |
| Consistent place for CSS styling | ❌ (MkDocs styling, custom transformers and Backstage CSS) | ✅ (can use CSS modules) |
| Indexing content | ✅ | ✅ (then build step would be mandatory) |
| `mkdocs-plugin-monorepo` | ✅ | ❌ (would need to be rewritten in React) |
| MkDocs `search` | ✅ | ❌ (would need to be rewritten in React) |
| MkDocs Admonition | ✅ | ✅ (straight-forward component) |
| GraphVIZ | ✅ (`markdown_inline_graphviz`) | ✅ (`graphviz-react`) |
| UML | ✅ (`plantuml_markdown`) | ⚠️ (`react-plantuml` but doesn't look very active) |
| Custom MkDocs plugins by our customers | ✅ | ❌ (would need to be rewritten in React) |

### Architectural overview

![TechdocsX-Architecture](https://user-images.githubusercontent.com/1255508/123109812-d6fd1000-d43b-11eb-95c3-fb42d165ede3.png)

### Proposed timeline and phases

Since this change eventually will affect every layer and interaction with TechDocs, a breakdown into smaller deliverables is necessary, along with integrating learnings from each step into the following ones.

#### Milestone 1: _Render subset of docs page on frontend_

We should implement a feature flag with which we can toggle rendering one aspect of the TechDocs page entirely on the client. This can be for example the table of contents. Following technical capabilities need to prepared:

- During the build step, copy raw docs (e.g. `mkdocs.yml` for the ToC) to the static output folder
- In the TechDocs backend, serve either raw YAML file, or parse and return `nav` property of `mkdocs.yml` in a new endpoint
- In the TechDocs reader, consume `nav` and render ToC using Material components, or existing Backstage components

For a PoC, see branch https://github.com/backstage/backstage/tree/techdocsx/milestone-0-poc

#### Milestone 2: _Render full docs page on frontend_

Achieve feature parity for e.g. code embeds, GraphVIZ, PlantUML, inline docs, search etc.

#### Milestone 3: _Add possibility to pre-build MDX assets_

With the same feature flag, enable processing MDX files during build time and persist in the same storage as now

#### Milestone 4: _Deprecate and remove MkDocs integration_

Start by using the TechdocsX approach by default and provide an opt-out via feature flag. Define a grace period, and eventually remove all MkDocs related code after that.

---

### Open questions

#### _Do we actually need to persist dynamic assets?_

The performance overhead of the MDX runtime might be negligible, needs to be confirmed by measurements (definitely orders of magnitudes faster than building with MkDocs).

Also, if indexing can be performed on MD(X) files, it might be possible to completely eliminate this part of the architecture. Needing input from users of Backstage

**Comment by freben** [comment]
@ottosichert Hey! Cool that you are looking into this. However, if I understand the proposal correctly, it sounds like it differs from what this issue discussed. I'm totally open to being wrong about that of course :) But, I wanted to ask, might it make sense to perhaps open a new issue instead, which is more targeted and thoroughly explaining the precise details and client-only aspects of what you wanted to do? Then that new issue could reference this old one as background info.

**Comment by ottosichert** [comment]
Hey @freben, thanks for your thoughts!

The originally proposed RFC and my follow-up should align on the general technical ideas. What I added on top was the possibility to have a purely client-side rendered flow for e.g. local development.

In production we would have a similar data flow as today, just that the number of layers is reduced. So we still process MD files at build time, just that everything behind the scenes then stays in JS land and is less complex.

The reason I think the client-side part should be part of this RFC is that it helps tackling this architectural change with smaller steps. So it could in theory be completely dropped (although it's a nice thing to have 😄)

**Comment by andrewthauer** [comment]
In general, I think this could streamline the experience, but is likely a very big undertaking. It sounds like this would be an entire docs editing & rendering platform (parts of it being from the ground up). I can't say I would miss the python specific markdown oddities that come with MkDocs though.

I would not want to rely on `mdx` for any standard/common markdown features that are handled in CommonMark / GH Flavoured, etc. markdown. Generally agreed upon markdown syntax standards should work as is, without introducing anything custom. Having a bring your own markdown docs and have it working great out of the box would be tenant for any solution imo.

One feature that is very important for us is using custom code fences and diagrams. PlantUML being the biggest. I'm pretty sure all PlantUML libraries (including ones for python, etc) are all just wrappers over the official Java based PlantUML (which actually uses `graphviz` under the covers). So rendering this on the fly in a browser isn't possible without a server. There is an offical PlantUML server that can be hosted + [Kroki](https://kroki.io/) does something similar, to proxy the image generation through a REST call. So this could be a limiting factor that requires some sort of server side dependencies.

In short though, I'm not attached to `MkDocs` and am generally supportive of removing the extra layers and steps required to get docs from writing to reading  (and back) easier. If the suggested solution makes the process simpler and handles the diagramming + not forcing you to use MDX for common things, then I think it's worth exploring more.

**Comment by ottosichert** [comment]
There are existing React components capable of rendering Graphviz Dot diagrams, e.g. [`graphviz-react`](https://www.npmjs.com/package/graphviz-react). Also directly available as `remark` plugin which fits into MDX: https://www.npmjs.com/package/@shedali/remark-graphviz

Similarly for PlantUML: https://www.npmjs.com/package/@akebifiky/remark-simple-plantuml

I would aim towards an integrated solution and to avoid external services, both SaaS and local rendering servers. I will investigate if the alternatives above are a fit for real-world diagrams.

MDX is a superset of CommonMark (https://github.com/mdx-js/specification#mdx). From what I can tell however there are differences in how incorrect HTML is being parsed, and obviously that JSX and `import/export` gets a special treatment within the superset of features.

**Comment by andrewthauer** [comment]
> There are existing React components capable of rendering Graphviz Dot diagrams, e.g. graphviz-react. Also directly available as remark plugin which fits into MDX: https://www.npmjs.com/package/@shedali/remark-graphviz

Yep, there is `d3-graphviz` which is what it's using. However, it wouldn't be feature complete afaik (supporting `dot` format only I believe). However, maybe that's not an issue in most cases.

> Similarly for PlantUML: https://www.npmjs.com/package/@akebifiky/remark-simple-plantuml

This also uses the PlantUML server proxy (see [plugin options](https://github.com/akebifiky/remark-simple-plantuml#plugin-options)). PlantUML is quite complex so I can't see it being possible to replicate 100% in the browser (or node). So some sort of server will be required to replicate this.

> MDX is a superset of CommonMark (https://github.com/mdx-js/specification#mdx). From what I can tell however there are differences in how incorrect HTML is being parsed, and obviously that JSX and import/export gets a special treatment within the superset of features.

Yep, for sure. Just saying that I would only venture into MDX if needed.

**Comment by ottosichert** [comment]
> This also uses the PlantUML server proxy

I indeed missed that! Then I see your point, it might be possible to use something like https://github.com/markushedvall/node-plantuml to generate the images at build time (with a local server), but client-side runtime only won't work easily.

> I would only venture into MDX if needed

I would argue this is a very popular and justified standpoint. To connect some dots and give an example, issue https://github.com/backstage/backstage/issues/4123 mentions supporting custom JS or support for a given library in general. In that case it would be possible to create a Backstage plugin that exports a component to TechDocs providing a Mermaid rendering component, e.g. using https://www.npmjs.com/package/react-mermaid2. Then, to make configuration explicit, the Mermaid component should be registered in some configuration file (alternatively auto-discover available TechDocs plugins, will add this as open question to my follow-up above).

In your MD file:

```
# Docs with Mermaid

Lorem ipsum dolor sit amet

<Mermaid>
graph LR
  a0((Third-party data))-- extracted -->a1[Event]
  a1-- transformed -->a2[Message]
  a2-- loaded -->a3[Record]
</Mermaid>
```

Thus the minimal use of JSX here is almost identical to HTML.

With `remark` it would also be possible to use an own syntax, for example hijacking existing code blocks and registering `mermaid` as a language, so e.g.:

    # Docs with Mermaid
    
    Lorem ipsum dolor sit amet
    
    ```mermaid
    graph LR
      a0((Third-party data))-- extracted -->a1[Event]
      a1-- transformed -->a2[Message]
      a2-- loaded -->a3[Record]
    ```

This should provide a clean API and flexibility to plugin creators to manage both simple plugins and more configurable ones with JSX. All in all, by providing custom components via plugins we can avoid the nightmare of including external, DOM-mutating JS assets in a React-rendered page.

**Comment by andrewthauer** [comment]
I wouldn't see any issues if both syntaxes are supported for this (MDX & old school code fenses). I've used `remark` in the past, and it's quite powerful with a very extensive eco-eystem. So being able tapping into that would be ideal.

It does feel a little like this proposal is overlapping with what `gatsby` and `next.js` provide though. Prior to adopting `Mkdocs`, I looked at using [Docz](https://www.docz.site/) which provided a lot of similar features built on top of `gatsby`. However, I found it required much more setup, whereas `mkdocs` was simple to run and built (literally a mkdocs.yml file and your off to the races). I'm hoping whatever solution is chosen the same is true. I'd love to be able to point to a directory of markdown files stored somewhere and have it work as a site.

On a side note, we are currently using the Mkdocs [awesome-pages plugin](https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin) which allows for navigation customization with `.pages` sidecar files. We've been able to avoid having a central `nav` tree and dynamically generate the mkdocs.yml on the fly so it's not required to exist.

**Comment by IuliiaBondarieva** [comment]
It would be really great to have this. Is there any progress in this direction?

**Comment by menosprezzi** [comment]
Soo nice! The
Also, it will be possible and easy to embed things from other solutions such as Excalidraw https://docs.excalidraw.com/docs/@excalidraw/excalidraw/integration#module-bundler

Today, we have to create a custom element and allow it from techdocs config. Thats the only way to achieve this.

The work in https://github.com/backstage/backstage/pull/3842 looks promising, even though is focused on client-side rendering.
